server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/RestoreInProgressAllocationDecider.java

```diff
@@ -46,25 +45,24 @@ public Decision canAllocate(final ShardRouting shardRouting, final RoutingAlloca
             return allocation.decision(Decision.YES, NAME, "ignored as shard is not being recovered from a snapshot");
         }
 
-        final Snapshot snapshot = ((RecoverySource.SnapshotRecoverySource) recoverySource).snapshot();
+        RecoverySource.SnapshotRecoverySource source = (RecoverySource.SnapshotRecoverySource) recoverySource;
         final RestoreInProgress restoresInProgress = allocation.custom(RestoreInProgress.TYPE);
 
         if (restoresInProgress != null) {
-            for (RestoreInProgress.Entry restoreInProgress : restoresInProgress.entries()) {
-                if (restoreInProgress.snapshot().equals(snapshot)) {
-                    RestoreInProgress.ShardRestoreStatus shardRestoreStatus = restoreInProgress.shards().get(shardRouting.shardId());
-                    if (shardRestoreStatus != null && shardRestoreStatus.state().completed() == false) {
-                        assert shardRestoreStatus.state() != RestoreInProgress.State.SUCCESS : "expected shard [" + shardRouting
-                            + "] to be in initializing state but got [" + shardRestoreStatus.state() + "]";
-                        return allocation.decision(Decision.YES, NAME, "shard is currently being restored");
-                    }
-                    break;
+            RestoreInProgress.Entry restoreInProgress = restoresInProgress.entries().get(source.restoreUUID());
```

Comment:
I added the new restore operation uuid to instances of `RecoverySource.SnapshotRecoverySource` => now we have a simple map lookup on the restore entries map here instead of the loop.

Reply:
We still need to expose the map since we do still loop over all entries or do an empty check in some spots.
Since we do the same (exposing the map) in e.g. `org.elasticsearch.cluster.RestoreInProgress.Entry#shards` it feels somewhat redundant to me but I don't have a strong opinion on it either.

Reply:
your call. With my suggestion to make `RestoreInProgress` an `Iterable<RestoreInProgress.Entry>` for easier iteration, we only need a getter and an isEmpty method.

Reply:
I like that (iterable + getter) => Will do :)